src/include/slab_containers.h

```diff
@@ -0,0 +1,728 @@
+// are larger (or smaller) than 'slabSize' nodes.
+//
+template<pool_index_t pool_ix,typename T,size_t stackSize, size_t slabSize>
+class slab_allocator : public pool_slab_allocator<pool_ix,T> {
+  struct slab_t;
+  struct slabHead_t {
+    slabHead_t *prev;
+    slabHead_t *next;
+  };
+  //
+  // Each slot has a pointer to it's containing slab PLUS either one Object OR if it's free exactly one pointer 
+  // in a per-slab freelist.
+  // Since this is raw memory, we don't want to declare something of type "T" to avoid
+  // accidental constructor/destructor calls.
+  //
+  struct slot_t {
+    slab_t *slab; // Pointer to my slab, NULL for slots on the stack (Technically, this could be an index but I'm lazy :))
+    enum { OBJECT_IN_POINTERS = (sizeof(T) + sizeof(void *) - 1) / sizeof(void *) };
+    slot_t *storage[OBJECT_IN_POINTERS]; // Either one object OR index[0] points to next element in free list for this slab
+  };
+   //
+   // Each slab has a freelist of objects within the slab and the size of that list
+   // The size is needed to cheaply determine when all of the slots within a slab are unused
+   // so that we can free the slab itself    
+   //
+   struct slab_t {
+      slabHead_t slabHead;  // membership of slab in list of slabs with free elements
+      uint32_t size;    // # of allocated slots in this slab
+      uint32_t freeSlots;   // # of free slots, i.e., size of freelist of slots within this slab
+      slot_t *freeHead;     // Head of list of freeslots OR NULL if none
+      slot_t slot[0];       // slots
+   };
+   slab_t *slabHeadToSlab(slabHead_t *head) { return reinterpret_cast<slab_t *>(head); }
+   //
+   // Exactly one of these as part of this container (the slab on the stack :))
+   //
+   struct stackSlab_t : public slab_t {
+      slot_t stackSlot[stackSize]; // Allows the compiler to get the right size :)
+   };
+   //
+   // Initialize a new slab, remember, "T" might not be correct use the stored sizes.
+   //
+   void initSlab(slab_t *slab,size_t sz) {
+      slab->size = sz;
+      slab->freeSlots = 0; // Pretend that it was completely allocated before :)
+      slab->freeHead = NULL;
+      slab->slabHead.next = NULL;
+      slab->slabHead.prev = NULL;
+      this->slab_allocate(sz,trueSlotSize,sizeof(slab_t));
+      char *raw = reinterpret_cast<char *>(slab->slot);
+      for (size_t i = 0; i < sz; ++i) {
+         slot_t *slot = reinterpret_cast<slot_t *>(raw);
+         slot->slab = slab;
+         freeslot(slot,false);
+         raw += trueSlotSize;
+      }
+   }
+   //
+   // Free a slot, the "freeEmpty" parameter indicates if this slab should be freed if it's emptied
+   // 
+   void freeslot(slot_t *s, bool freeEmpty) {
+      slab_t *slab = s->slab;
+      //
+      // Put this slot onto the per-slab freelist
+      //
+      s->storage[0] = slab->freeHead;
+      slab->freeHead = s;
+      slab->freeSlots++;
+      ++freeSlotCount;
+      this->slab_item_free(trueSlotSize);
```

Comment:
Houston, we have a problem! The `slab_allocator` is tightly glued to the atomical accounting machinery of `mempool`. When applied to `WeightedPriorityQueue` (generates nasty traffic from `tcmalloc`'s POV), a `lock`-driven regression pops up:
```
Samples: 8K of event 'cycles:pp', Event count (approx.): 2472199698                                                                                                                                           
  Children      Self  Command        Shared O  Symbol                                                                                                                                                        ◆
-    4,70%     4,03%  msgr-worker-0  ceph-osd  [.] WeightedPriorityQueue<OpQueueItem, unsigned long>::Queue::insert                                                                                          ▒
   - 4,02% __clone                                                                                                                                                                                           ▒
        start_thread                                                                                                                                                                                         ▒
        execute_native_thread_routine                                                                                                                                                                        ▒
        std::_Function_handler<void (), NetworkStack::add_thread(unsigned int)::{lambda()#1}>::_M_invoke                                                                                                     ▒
        EventCenter::process_events                                                                                                                                                                          ▒
        AsyncConnection::process                                                                                                                                                                             ▒
        DispatchQueue::fast_dispatch                                                                                                                                                                         ▒
        OSD::ms_fast_dispatch                                                                                                                                                                                ▒
        OSD::enqueue_op                                                                                                                                                                                      ▒
        OSD::ShardedOpWQ::_enqueue                                                                                                                                                                           ▒
        WeightedPriorityQueue<OpQueueItem, unsigned long>::Queue::insert                                                                                                                                     ▒
   + 0,66% WeightedPriorityQueue<OpQueueItem, unsigned long>::Queue::insert                                                                                                                                  ▒
```

```
       │     _ZN7mempool14slab_allocatorILNS_12pool_index_tE19ESt10_List_nodeIN21WeightedPriorityQueueI11OpQueueItemmE8ListPairEELm32ELm32EE8freeslotEPNS8_6slot_tEb():                                       
       │           slab_t *slab = s->slab;                                                                                                                                                                    
       │           //                                                                                                                                                                                         
       │           // Put this slot onto the per-slab freelist                                                                                                                                                
       │           //                                                                                                                                                                                         
       │           s->storage[0] = slab->freeHead;                                                                                                                                                            
       │           slab->freeHead = s;                                                                                                                                                                        
       │       mov    %rax,0x80(%rbp)                                                                                                                                                                         
       │     _ZNSt13__atomic_baseIlEpLEl():                                                                                                                                                                   
 64,76 │       lock   add    %r8,(%rdx,%r13,1)                                                                                                                                                                
 14,76 │       lock   addq   $0x1,(%rdx,%rcx,1)                                                                                                                                                               
       │     _ZN7mempool19pool_slab_allocatorILNS_12pool_index_tE19ESt10_List_nodeIN21WeightedPriorityQueueI11OpQueueItemmE8ListPairEEE14slab_item_freeEm():                                                  
       │         shard->free_bytes += sizeof_T;                                                                                                                                                               
       │         shard->free_items += 1;                                                                                                                                                                      
```

Possibilities:
  1. decouple the `slab_allocator` from `mempool` (let it be used as non-accounted replacement for STL containers).
  2. Pacify the `mempool` with TLS in a way similar to `perf_counters_t`. As the stuff inlines nicely, we might have the same problems elsewhere. TODO: instrument `mempool` and profile.
  3. ?